### PR TITLE
Closes #3241: Add DB update for fixing langcode on migrated menu links and add DefaultLangcode migrate plugin.

### LINKED ISSUE #3241: Menu link parent can be lost after saving a page as draft and then published [Placeholder]
## Problem/Motivation
Why the issue was filed, steps to reproduce the problem, etc.

## Proposed resolution
Description of the proposed solution, the rationale behind it, and workarounds for people who cannot use the patch.

##  Screenshots
If applicable, add screenshots to help explain your problem. (Need a tool to share screenshots or screencasts as URLs? Try [monosnap](https://monosnap.com)

### PR DESCRIPTION
**Pull request by joeparsons**
## Description

- Adds DB update to fix menu link content entities that had their langcode value set to `LanguageInterface::LANGCODE_NOT_SPECIFIED` (`und`) as part of a migration to address the strange behavior we've seen with the parent values for menu links changing unexpectedly under certain content moderation conditions
- Adds a new migrate process plugin (`az_default_langcode`) for use in content entity migrations
- Updates various migrations in `az_migration` (including `az_menu_links`) to use new process plugin.

## Related issues

Closes #3241

## How to test

This PR must be tested on a site that was migrated using the az_migration module and which still has menu links that are set to the `und` (Undetermined) language code.

We tested this PR on one of our sites that exhibited the menu link problem. The database update resulted in the following output:

```
Notice: >  [notice] Update started: az_quickstart_update_1020901
Notice: >  [notice] Updated langcode on 25 menu links.
Notice: >  [notice] Updated langcode on 50 menu links.
Notice: >  [notice] Updated langcode on 75 menu links.
Notice: >  [notice] Updated langcode on 100 menu links.
Notice: >  [notice] Updated langcode on 125 menu links.
Notice: >  [notice] Updated langcode on 150 menu links.
Notice: >  [notice] Updated langcode on 159 menu links.
Notice: >  [notice] Update completed: az_quickstart_update_1020901
```

After the update, no menu links remained with the `und` langcode. Also, in testing a page that exhibited the problem before when saving it as Draft and then saving it again as Published, the inadvertent change to the menu link parent could not be reproduced.

### Arizona Quickstart (install profile, custom modules, custom theme)
- **Patch release changes**
   - [x] Bug fix
   - [ ] Accessibility, performance, or security improvement
   - [ ] Critical institutional link or brand change
   - [ ] Adding experimental module
   - [ ] Update experimental module
- **Minor release changes**
   - [x] New feature
   - [ ] Breaking or visual change to existing behavior
   - [ ] Upgrade experimental module to stable
   - [ ] Enable existing module by default or database update
   - [ ] Non-critical brand change
   - [ ] New internal API or API improvement with backwards compatibility
   - [ ] Risky or disruptive cleanup to comply with coding standards
   - [ ] High-risk or disruptive change (requires upgrade path, risks regression, etc.)
- **Other or unknown**
   - [ ] Other or unknown

### Drupal core
- **Patch release changes**
   - [ ] Security update
   - [ ] Patch level release (non-security bug-fix release)
   - [ ] Patch removal that's no longer necessary
- **Minor release changes**
   - [ ] Major or minor level update
- **Other or unknown**
   - [ ] Other or unknown

### Drupal contrib projects
- **Patch release changes**
   - [ ] Security update
   - [ ] Patch or minor level update
   - [ ] Add new module
   - [ ] Patch removal that's no longer necessary
- **Minor release changes**
   - [ ] Major level update
- **Other or unknown**
   - [ ] Other or unknown

## Checklist

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: az_quickstart.install [hunk 1]
```diff
@@ -195,3 +195,31 @@ function az_quickstart_update_1020802() {
...
 }
+
+/**
+ * Update langcode for migrated menu links to be the site's default language.
+ */
+function az_quickstart_update_1020901(&$sandbox) {
+  if (!isset($sandbox['progress'])) {
+    $sandbox['progress'] = 0;
+    $sandbox['max'] = \Drupal::entityQuery('menu_link_content')
+      ->condition('langcode', 'und')
```
Comment (joeparsons) [inline az_quickstart.install:206]:
```suggestion
      ->accessCheck(FALSE)
      ->condition('langcode', 'und')
```

File: az_quickstart.install [hunk 2]
```diff
@@ -195,3 +195,31 @@ function az_quickstart_update_1020802() {
...
+      ->count()
+      ->execute();
+  }
+  $batch_size = 25;
+  $default_langcode = \Drupal::languageManager()->getDefaultLanguage()->getId();
+
+  $ids = \Drupal::entityQuery('menu_link_content')
+    ->condition('langcode', 'und')
```
Comment (joeparsons) [inline az_quickstart.install:214]:
```suggestion
    ->accessCheck(FALSE)
    ->condition('langcode', 'und')
```

File: az_quickstart.install [hunk 3]
```diff
@@ -195,3 +195,33 @@ function az_quickstart_update_1020802() {
 function az_quickstart_update_1020803() {
   \Drupal::service('module_installer')->uninstall(['media_library_theme_reset']);
 }
+
+/**
+ * Update langcode for migrated menu links to be the site's default language.
+ */
+function az_quickstart_update_1020901(&$sandbox) {
+  if (!isset($sandbox['progress'])) {
```
Comment (joeparsons) [inline az_quickstart.install:203]:
```suggestion
  if (!isset($sandbox['max'])) {
```

File: az_quickstart.install [hunk 4]
```diff
@@ -195,3 +195,33 @@ function az_quickstart_update_1020802() {
...
+      ->count()
+      ->execute();
+  }
+  $batch_size = 25;
+  $default_langcode = \Drupal::languageManager()->getDefaultLanguage()->getId();
+
+  $ids = \Drupal::entityQuery('menu_link_content')
+    ->accessCheck(FALSE)
+    ->condition('langcode', 'und')
+    ->range($sandbox['progress'], $sandbox['progress'] + $batch_size)
```
Comment (joeparsons) [inline az_quickstart.install:217]:
```suggestion
    ->range($sandbox['progress'], $batch_size)
```

File: az_quickstart.install [hunk 5]
```diff
@@ -195,3 +195,33 @@ function az_quickstart_update_1020802() {
...
+    ->condition('langcode', 'und')
+    ->range($sandbox['progress'], $batch_size)
+    ->execute();
+  $menu_links = \Drupal::entityTypeManager()->getStorage('menu_link_content')->loadMultiple($ids);
+  foreach ($menu_links as $menu_link) {
+    $menu_link->langcode = $default_langcode;
+    $menu_link->save();
+    $sandbox['progress']++;
+  }
+  $sandbox['#finished'] = $sandbox['max'] === 0 ? 1 : $sandbox['progress'] / $sandbox['max'];
```
Comment (joeparsons) [inline az_quickstart.install:225]:
```suggestion
  $sandbox['#finished'] = empty($sandbox['max']) ? 1 : $sandbox['progress'] / $sandbox['max'];
```

File: az_quickstart.install [hunk 6]
```diff
@@ -195,3 +195,33 @@ function az_quickstart_update_1020802() {
 function az_quickstart_update_1020803() {
   \Drupal::service('module_installer')->uninstall(['media_library_theme_reset']);
 }
+
+/**
+ * Update langcode for migrated menu links to be the site's default language.
+ */
+function az_quickstart_update_1020901(&$sandbox) {
+  if (!isset($sandbox['max'])) {
```
Comment (joeparsons) [inline az_quickstart.install:203]:
```suggestion
  if (!isset($sandbox['progress'])) {
```

File: modules/custom/az_migration/src/Plugin/migrate/process/DefaultLangcode.php [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+   */
+  public function __construct(array $configuration, $plugin_id, $plugin_definition, LanguageManagerInterface $language_manager) {
+    parent::__construct($configuration, $plugin_id, $plugin_definition);
+    $this->languageManager = $language_manager;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function create(ContainerInterface $container, array $configuration, $plugin_id, $plugin_definition) {
```
Comment (bberndt-uaz) [inline modules/custom/az_migration/src/Plugin/migrate/process/DefaultLangcode.php:62]: Currently, the PR is producing an error in our migration tests:

> TypeError: Drupal\az_migration\Plugin\migrate\process\DefaultLangcode::__construct(): Argument #4 ($language_manager) must be of type Drupal\Core\Language\LanguageManagerInterface, Drupal\migrate\Plugin\Migration given, called in /src/az-quickstart-scaffolding/web/core/modules/migrate/src/Plugin/MigratePluginManager.php on line 60 in Drupal\az_migration\Plugin\migrate\process\DefaultLangcode->__construct() (line 54 of /src/az-quickstart-scaffolding/web/profiles/custom/az_quickstart/modules/custom/az_migration/src/Plugin/migrate/process/DefaultLangcode.php).
> [error]  TypeError: Drupal\az_migration\Plugin\migrate\process\DefaultLangcode::__construct(): Argument #4 ($language_manager) must be of type Drupal\Core\Language\LanguageManagerInterface, Drupal\migrate\Plugin\Migration given, called in /src/az-quickstart-scaffolding/web/core/modules/migrate/src/Plugin/MigratePluginManager.php on line 60 in Drupal\az_migration\Plugin\migrate\process\DefaultLangcode->__construct() (line 54 of /src/az-quickstart-scaffolding/web/profiles/custom/az_quickstart/modules/custom/az_migration/src/Plugin/migrate/process/DefaultLangcode.php) #0 /src/az-quickstart-scaffolding/web/core/modules/migrate/src/Plugin/MigratePluginManager.php(60): Drupal\az_migration\Plugin\migrate\process\DefaultLangcode->__construct()

Perhaps this change is needed?

```suggestion
  public static function create(ContainerInterface $container, array $configuration, $plugin_id, $plugin_definition, MigrationInterface $migration = NULL) {
```

Comment (joeparsons) [reply modules/custom/az_migration/src/Plugin/migrate/process/DefaultLangcode.php:62]: Thanks @bberndt-uaz

**Review by bberndt-uaz — APPROVED** [review]

**Review by tadean — APPROVED** [review]

**Pull request merged**